plugins/experimental/filter_body/filter_body.cc

```diff
@@ -0,0 +1,771 @@
+{
+  TSMLoc field_loc = TSMimeHdrFieldFind(bufp, hdr_loc, cond.name.c_str(), static_cast<int>(cond.name.length()));
+  if (field_loc == TS_NULL_MLOC) {
+    return false;
+  }
+
+  bool matched = false;
+  // Check all values of this header field
+  int num_values = TSMimeHdrFieldValuesCount(bufp, hdr_loc, field_loc);
+  for (int i = 0; i < num_values && !matched; ++i) {
+    int         value_len = 0;
+    const char *value     = TSMimeHdrFieldValueStringGet(bufp, hdr_loc, field_loc, i, &value_len);
+    if (value == nullptr) {
+      continue;
+    }
+
+    // Check if any pattern matches (OR logic within header)
+    for (const auto &pattern : cond.patterns) {
+      if (strcasestr_local(value, value_len, pattern.c_str(), pattern.length()) != nullptr) {
+        matched = true;
+        break;
+      }
+    }
+  }
+
+  TSHandleMLocRelease(bufp, hdr_loc, field_loc);
+  return matched;
+}
+
+// Check if ALL header conditions match (AND logic between headers)
+bool
+headers_match(const Rule &rule, TSMBuffer bufp, TSMLoc hdr_loc)
+{
+  for (const auto &cond : rule.headers) {
+    if (!header_condition_matches(cond, bufp, hdr_loc)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Search for body patterns in data (case-sensitive)
+// Returns the matched pattern or nullptr
+const std::string *
+search_body_patterns(const Rule &rule, const char *data, size_t data_len)
+{
+  for (const auto &pattern : rule.body_patterns) {
+    if (strstr_local(data, data_len, pattern.c_str(), pattern.length()) != nullptr) {
+      return &pattern;
+    }
+  }
+  return nullptr;
+}
+
+// Add header to the request/response
+void
+add_header_to_message(TSMBuffer bufp, TSMLoc hdr_loc, const std::string &name, const std::string &value)
+{
+  TSMLoc field_loc;
+  if (TSMimeHdrFieldCreateNamed(bufp, hdr_loc, name.c_str(), static_cast<int>(name.length()), &field_loc) != TS_SUCCESS) {
+    TSError("[%s] Failed to create header field: %s", PLUGIN_NAME, name.c_str());
+    return;
+  }
+
+  if (TSMimeHdrFieldValueStringSet(bufp, hdr_loc, field_loc, -1, value.c_str(), static_cast<int>(value.length())) != TS_SUCCESS) {
+    TSError("[%s] Failed to set header value: %s", PLUGIN_NAME, name.c_str());
+    TSHandleMLocRelease(bufp, hdr_loc, field_loc);
+    return;
+  }
+
+  if (TSMimeHdrFieldAppend(bufp, hdr_loc, field_loc) != TS_SUCCESS) {
+    TSError("[%s] Failed to append header field: %s", PLUGIN_NAME, name.c_str());
+  }
+
+  TSHandleMLocRelease(bufp, hdr_loc, field_loc);
+}
+
+// Execute actions for a matched rule
+void
+execute_actions(TransformData *data, const Rule *rule, const std::string *matched_pattern)
+{
+  if (rule->actions & ACTION_LOG) {
+    Dbg(dbg_ctl, "Matched rule: %s, pattern: %s", rule->name.c_str(), matched_pattern ? matched_pattern->c_str() : "unknown");
+  }
+
+  if ((rule->actions & ACTION_ADD_HEADER) && !data->headers_added) {
+    TSMBuffer bufp;
+    TSMLoc    hdr_loc;
+
+    // Add header to server request (proxy request going to origin)
```

Comment:
take a look at the direction to see if it is a request or a response and then add the header